src/riscv/sysv.S

```diff
@@ -0,0 +1,782 @@
+#define FFI_FLAG_BITS_X5 (5 * FFI_FLAG_BITS)
+#define FFI_FLAG_BITS_X6 (6 * FFI_FLAG_BITS)
+#define FFI_FLAG_BITS_X7 (7 * FFI_FLAG_BITS)
+
+# Stack pointer needs to be 16-byte aligned, so frame size is rounded up
+#define SIZEOF_FRAME (6 * FFI_SIZEOF_ARG)
+
+    .text
+    .align  2
+    .globl  ffi_call_asm
+    .type   ffi_call_asm, @function
+ffi_call_asm:
+    .cfi_startproc
+    
+    ### Prologue
+    
+    # a0 - ffi_prep_args pointer
+    # a1 - extended_cif pointer
+    # a2 - bytes
+    # a3 - flags
+    # a4 - rvalue
+    # a5 - function ptr
+    
+    add     sp, sp, -SIZEOF_FRAME   # move stack pointer by frame size
+                                    # must be 16-byte aligned 
+    
+    # stack ptr points to first argument on stack,
+    # but there should be no arguments on the stack
+    
+    .cfi_def_cfa_offset SIZEOF_FRAME
+    
+    REG_S   ra,      FFI_SIZEOF_ARG_X5(sp)   # save return address
+    REG_S   fp,      FFI_SIZEOF_ARG_X4(sp)   # save frame pointer
+    REG_S   flags,   FFI_SIZEOF_ARG_X3(sp)   # save flags
+    REG_S   fn,      FFI_SIZEOF_ARG_X2(sp)   # save function pointer
+    REG_S   rvalue,  FFI_SIZEOF_ARG_X1(sp)   # save return value pointer
+    
+    .cfi_offset 1, -8
+    .cfi_offset 8, -16
+    
+    add     fp, sp, SIZEOF_FRAME     # new frame pointer is old stack pointer
+    
+    .cfi_def_cfa 8, 0
+    
+    add     t4, callback, zero       # function ptr to prep_args
+    
+    # Here we're setting up our argspace and its size
+    
+    add     t0, bytes, 15      # make sure it is aligned 
+    andi    t0, t0,   -16      # to a 16 byte boundry
+
+thirtytwo:
+    sub     sp, sp, t0 # move the stack pointer to reflect the arg space
+    
+    # a0 is the stack with proper arg space allocated
+    add     a0, sp, zero
+    
+    # a1 is ecif
+    # a2 is bytes
+    # a3 is flags
+
+    jalr    t4  # call ffi_prep_args
+    
+    REG_L   t0, -FFI_SIZEOF_ARG_X3(fp)  # load the flags word
+    srli    t2, t0, 16                 # shift our return type into t2
+    
+    li      t1, ARG_MASK
+    and     t0, t0, t1                 # mask out the arg types into t0
+
+    # time to load the arguments for the call
+
+#ifndef __riscv_soft_float
+
+####################
+## SET ARGUMENT 0 ##
+####################
+
+set_arg0:
+    srli    t1, t0, FFI_FLAG_BITS_X0 # Shift to get the bits for this argument
+    andi    t1, t1, 3               # Mask out the bits for this argument
+    
+    # when its zero, it means its just a word-sized int/ptr
+    bne     t1, zero, set_arg0_float 
+    REG_L   a0, FFI_SIZEOF_ARG_X0(sp) # load argument
+    j       set_arg1
+    
+set_arg0_float:
+    addi    t1, t1, -2
+    
+    # when its zero, it means its just a word-sized float
+    bne     t1, zero, set_arg0_double
+    flw     fa0, FFI_SIZEOF_ARG_X0(sp) # load argument
+    j       set_arg1
+    
+set_arg0_double:
+    # otherwise it must be a double we're dealing with
+    fld     fa0, FFI_SIZEOF_ARG_X0(sp)
+
```

Comment:
Agreed; always setting FA[0-7] is safe.  One possible exception is that there are RISC-V implementations in the wild that trap and emulate all FP instructions, including FLD/FLW; it might be nice to have a single branch with options "set the integer registers" and "set all registers, at least one FP argument exists", but of course I have no hardware performance numbers.